### PR TITLE
FTS index and functions

### PR DESCRIPTION
**Pull request by vpetrovykh**

### CODE REVIEW

File: edb/pgsql/delta.py [hunk 1]
```diff
@@ -3263,19 +3265,54 @@ def create_index(cls, index, schema, context):
...
+                as_fragment=True,
+            )
+            kw_ir = kw_expr.irast
+            kw_sql_tree = compiler.compile_ir_to_sql_tree(
+                kw_ir.expr, singleton_mode=True)
+            sql = codegen.SQLSourceGenerator.to_source(kw_sql_tree)
+            # HACK: the compiled SQL is expected to have some unnecessary
+            # casts, strip casts to text as they mess with the requirement
+            # that index expressions are IMMUTABLE.
+            if sql.endswith('::text'):
```
Comment (msullivan) [inline edb/pgsql/delta.py:3300]: Hmmmmm this is very sketchy. What casts were running afoul of this?

File: edb/schema/indexes.py [hunk 2]
```diff
@@ -125,6 +129,23 @@ class Index(
         inheritable=False,
     )
 
+    params = so.SchemaField(
```
Comment (msullivan) [inline edb/schema/indexes.py:132]: Add a comment around here making clear when `params` appears and when `kwargs` does?

File: edb/schema/indexes.py [hunk 3]
```diff
@@ -185,6 +206,57 @@ def get_displayname_static(cls, name: sn.Name) -> str:
         shortname = cls.get_shortname_static(name)
         return shortname.name
 
+    def get_all_kwargs(
+        self,
+        schema: s_schema.Schema,
+    ) -> s_expr.ExpressionDict:
+        kwargs = s_expr.ExpressionDict()
+        all_kw = self.__class__.get_field('kwargs').merge_fn(
```
Comment (msullivan) [inline edb/schema/indexes.py:214]: Does just `self.get_field` not work? (Or does that get resolved to something else?)
I'd probably go with `type(self)` over `self.__class__`

File: edb/schema/indexes.py [hunk 4]
```diff
@@ -367,16 +458,56 @@ def _cmd_from_ast(
...
+    ) -> Optional[qlast.DDLOperation]:
+        astnode = super()._get_ast(schema, context, parent_node=parent_node)
+
+        kwargs = self.get_resolved_attribute_value(
+            'kwargs',
+            schema=schema,
+            context=context,
+        )
+        if kwargs:
+            astnode.kwargs = dict()
```
Comment (msullivan) [inline edb/schema/indexes.py:476]: Might be cleaner to do a dict comprehension

File: edb/schema/indexes.py [hunk 5]
```diff
@@ -125,6 +129,23 @@ class Index(
...
+        coerce=True,
+        compcoef=0.4,
+        default=so.DEFAULT_CONSTRUCTOR,
+        inheritable=True,
+    )
+
+    kwargs = so.SchemaField(
+        s_expr.ExpressionDict,
+        coerce=True,
+        compcoef=0.4,
```
Comment (msullivan) [inline edb/schema/indexes.py:143]: Does `ALTER` work for any of this yet? If not maybe compcoef should be 0?

File: edb/pgsql/metaschema.py [hunk 6]
```diff
@@ -4134,6 +4134,130 @@ def __init__(self) -> None:
         )
 
 
+class FTSParseQueryFunction(dbops.Function):
```
Comment (msullivan) [inline edb/pgsql/metaschema.py:4137]: Is this going to be fast enough? I guess there's no way around having to do this kind of computation, if we're doing a string input query.

Comment (msullivan) [reply edb/pgsql/metaschema.py:4137]: I'd bet that looping instead of doing recursion would help, though?

File: edb/pgsql/delta.py [hunk 7]
```diff
@@ -3263,19 +3265,57 @@ def create_index(cls, index, schema, context):
...
+                kw_expr = expr.ensure_compiled(
+                    schema=schema,
+                    options=kw_options,
+                    as_fragment=True,
+                )
+                kw_ir = kw_expr.irast
+                kw_sql_tree = compiler.compile_ir_to_sql_tree(
+                    kw_ir.expr, singleton_mode=True)
+                sql = codegen.SQLSourceGenerator.to_source(kw_sql_tree)
+                # HACK: the compiled SQL is expected to have some unnecessary
```
Comment (msullivan) [inline edb/pgsql/delta.py:3300]: Hmmmmm this is very sketchy. What casts were running afoul of this?

File: edb/pgsql/delta.py [hunk 8]
```diff
@@ -3263,19 +3265,57 @@ def create_index(cls, index, schema, context):
...
+            # constants)
+            # These are expected to be constants, so we don't have anchors,
+            # path prefixes, etc.
+            kw_options = qlcompiler.CompilerOptions(
+                modaliases=context.modaliases,
+                schema_object_context=cls.get_schema_metaclass(),
+                apply_query_rewrites=False,
+            )
+            for name, expr in kwargs.items():
+                # XXX: origin messes up compilation, but by this point we
```
Comment (msullivan) [inline edb/pgsql/delta.py:3288]: How does origin mess up the compilation?

File: tests/test_indexes.py [hunk 9]
```diff
@@ -231,3 +231,34 @@ async def test_index_05(self):
                 };
                 """
             )
+
```
Comment (msullivan) [inline tests/test_indexes.py:234]: We also need some tests for creating new abstract indexes via extending, right?
(Though that's definitely lower priority than FTS, so if it turns out to be pretty busted and tricky to get working we might be able to punt on that for now)

File: edb/edgeql/parser/grammar/commondl.py [hunk 10]
```diff
@@ -639,6 +657,20 @@ def _process_arguments(self, arguments):
 
         return kwargs
 
+    def _process_params_or_kwargs(self, bases, arguments):
+        params = []
+        kwargs = dict()
+
+        # If the definition is extending anotehr abstract index, then we
```
Comment (1st1) [inline edb/edgeql/parser/grammar/commondl.py:664]:
```suggestion
        # If the definition is extending another abstract index, then we
```

File: edb/pgsql/delta.py (hunk 7 above)
Comment (1st1) [reply edb/pgsql/delta.py:3300]: +1.

File: edb/schema/reflection/writer.py [hunk 11]
```diff
@@ -383,6 +383,46 @@ def _build_object_mutation_shape(
 
             assignments.append(f'{ns}__internal := {shadow_target_expr}')
 
+        elif ftype is sr_struct.FieldType.EXPR_DICT:
+            target_expr = f'''
```
[the remark this answers is not in the document]
Comment (vpetrovykh) [reply edb/schema/reflection/writer.py:387]: Updated

File: edb/schema/indexes.py (hunk 5 above)
Comment (vpetrovykh) [reply edb/schema/indexes.py]: Set to 0

**Review by msullivan — APPROVED** [review]
OK, my understanding is that there are still a number of FTS changes planned, but the index side of things is mostly in place (except for ALTER?), and can be merged?
And then we'll follow up some more with FTS-specific PRs?

**Comment by vpetrovykh** [comment]
Yeah, that's the plan.

**Pull request merged**